File: src/GarminFilter.Api/Models/AppViewModel.cs
using GarminFilter.Domain.App.Aggregates;
using GarminFilter.Domain.App.ValueObjects;

namespace GarminFilter.Api.Models;

public class AppViewModel
{
	public AppViewModel(IGarminApp app)
	{
		Id = app.Id;
		Name = app.Name;
		Type = app.TypeId;
		ImageUrl = $"/app/{app.IconFileId}";
		Url = $"https://apps.garmin.com/apps/{app.Id}";
		IsPaid = app.IsPaid;
		Permissions = app.Permissions
			.Select(permission => new AppPermissionViewModel(permission))
			.OrderBy(permission => permission.Description)
			.ToList();
		Developer = app.DeveloperName;
	}

	public string Developer { get; }


	public List<AppPermissionViewModel> Permissions { get; }

	public bool IsPaid { get; }

	public string Url { get; }

	public string ImageUrl { get; }

	public AppType Type { get; }

	public AppId Id { get; }
	public string Name { get; }
}

File: flutter-frontend/lib/models/app_view_model.dart
import 'permission.dart';

class AppViewModel {
  final List<Permission> permissions;
  final bool isPaid;
  final String url;
  final String imageUrl;
  final String type;
  final String id;
  final String name;
  final String developer;

  AppViewModel({
    required this.permissions,
    required this.isPaid,
    required this.url,
    required this.imageUrl,
    required this.type,
    required this.id,
    required this.name,
    required this.developer,
  });

  factory AppViewModel.fromJson(Map<String, dynamic> json) {
    return AppViewModel(
      permissions: json['permissions'] != null 
          ? (json['permissions'] as List).map((p) => Permission.fromJson(p)).toList()
          : [],
      isPaid: json['isPaid'] as bool,
      url: json['url'] as String,
      imageUrl: json['imageUrl'] as String,
      type: json['type'] as String,
      id: json['id'] as String,
      name: json['name'] as String,
      developer: json['developer'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'permissions': permissions.map((p) => p.toJson()).toList(),
      'isPaid': isPaid,
      'url': url,
      'imageUrl': imageUrl,
      'type': type,
      'id': id,
      'name': name,
      'developer': developer,
    };
  }
}
